### PR TITLE
Codegen array parse matrices fixes

### PR DESCRIPTION
**Pull request by Upabjojr**
#### Release Notes

* codegen
  * Fix bugs to the parser of matrix expressions in order to build CodegenArray objects correctly.

### CODE REVIEW

File: sympy/codegen/array_utils.py [hunk 1]
```diff
@@ -435,22 +435,6 @@ def _get_contraction_links(self):
         args, dlinks = _get_contraction_links([self], self.subranks, *self.contraction_indices)
         return dlinks
 
-    @staticmethod
-    def from_MatMul(expr):
```
Comment (sylee957) [inline sympy/codegen/array_utils.py:439]: Should this follow the deprecation cycle?

File: sympy/codegen/array_utils.py [hunk 2]
```diff
@@ -1026,22 +1016,32 @@ def _codegen_array_parse(expr):
             args[i] = CodegenArrayPermuteDims(args[i], permutation)
         return CodegenArrayElementwiseAdd(*args), index0
     return expr, ()
-    raise NotImplementedError("could not recognize expression %s" % expr)
 
 
-def _parse_matrix_expression(expr):
+def parse_matrix_expression(expr: MatrixExpr) -> Basic:
+    return _parse_matrix_expression(expr)
```
Comment (asmeurer) [inline sympy/codegen/array_utils.py:1022]: What's the purpose of this indirection? Is it just for backwards compatibility for the name `_parse_matrix_exception`?

File: sympy/codegen/array_utils.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/codegen/array_utils.py:439]: No, it wasn't working properly. I doubt people were using it.

File: sympy/codegen/array_utils.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/codegen/array_utils.py:1022]: I just wanted to avoid exposing a recurrent function publicly. Would you prefer to change it?

Comment (asmeurer) [reply sympy/codegen/array_utils.py:1022]: I see. Usually this is done if there is some preprocessing logic that should only be done once. I'm not sure if it's beneficial to have it done trivially like it is here.

**Pull request merged**